flink-connectors/flink-connector-files/src/test/java/org/apache/flink/connector/file/table/FileSystemOutputFormatTest.java

```diff
@@ -104,15 +113,101 @@ void testNonPartition() throws Exception {
                 .containsExactly("a1,1,p1\n" + "a2,2,p1\n" + "a2,2,p2\n" + "a3,3,p1\n");
     }
 
+    @ParameterizedTest
+    @ValueSource(booleans = {true, false})
+    void testStagingDirBehavior(boolean shareStagingDir) throws Exception {
```

Comment:
Generally speaking, unit tests should be as small as possible to focus on the specific aspect of the class contract (in this case `FileSystemTableSink#toStagingPath` not returning the same folder twice. This test adds quite a bit of extra context here which might make it harder to grasp the intention for code readers. What's your opinion on that?

Reply:
> Generally speaking, unit tests should be as small as possible to focus on the specific aspect of the class contract (in this case `FileSystemTableSink#toStagingPath` not returning the same folder twice. This test adds quite a bit of extra context here which might make it harder to grasp the intention for code readers. What's your opinion on that?

+1 with your perspective on unit testing. As I previously mentioned, in this case, constructing a UT is quite challenging, which subtly suggests to me that there may be some design issues with the current code. Take the staging dir, for example, if it needs to be passed from the `FileSystemTableSink` to the `OutputFormat` at compile-time, then a proper validation should be done here. Conversely, perhaps the `OutputFormat` could decide on the stagingDir itself, which would make the testing more targeted.

Reply:
I guess we're getting closer here. :-) I'm not that familiar with this part of the code. But as said earlier, a complicated verbose unit test is an indication for a non-optimal code structure. Also checking the interfaces' JavaDocs backs your proposal:

For `DynamicTableSink` (which is what `FileSystemTableSink` implements):
> Sink of a dynamic table to an external storage system.

For `OutputFormat` (which is what `FileSystemOutputFormat` implements):
> The base interface for outputs that consumes records. The output format describes how to store the final records, for example in a file.

The key sentence here is `The output format describes how to store the final records, for example in a file.`. The "how" should include the "where", i.e. the `OutputFormat` implementation should be in charge of creating the staging folder and, therefore, would be also in charge of create a folder that is not used by any other `OutputFormat` instance.